core/src/views/mod.rs

```diff
@@ -0,0 +1,288 @@
+use crate::{
+    messages::{
+        agent::{BackendState, DroneState},
+        state::StateUpdate,
+    },
+    types::{BackendId, ClusterName, DroneId},
+};
+use std::{
+    collections::{BTreeMap, HashMap},
+    net::IpAddr,
+};
+use time::OffsetDateTime;
+
+pub mod replica;
+
+#[derive(Default)]
+pub struct DroneView {
+    /// The most recent heartbeat we've received for this drone.
+    pub last_heartbeat: Option<(DroneState, OffsetDateTime)>,
+    pub ip: Option<IpAddr>,
+    pub version: Option<String>,
+
+    /// A map of backends relevant to scheduling or routing.
+    /// When backends terminate, they are removed from this map.
+    pub backends: HashMap<BackendId, BackendState>,
+}
+
+impl DroneView {
+    pub fn state(&self) -> Option<DroneState> {
+        let (last_drone_state, _timestamp) = self.last_heartbeat?;
+        Some(last_drone_state)
+    }
+
+    pub fn num_backends(&self) -> usize {
+        self.backends.len()
+    }
+}
+
+#[derive(Default)]
+pub struct ClusterView {
+    /// Maps active backends to the drone they run on.
+    pub routes: BTreeMap<BackendId, DroneId>,
```

Comment:
why Btrees instead of hashmaps?

Reply:
I wanted to preserve the (recent) behaviour that `list-dns` and `list-drones` return drones/backends in alphabetical order.

I suppose it would be slightly more performant to use `HashMap` and sort only in the CLI after they are retrieved, but this is simple for now and I don't think it will be a noticeable performance difference.

Reply:
my worry here is that lookups are no longer O(1), which shouldn't be an issue, but is another thing to think about.

api doesn't really change though if we find we need to go to a hashmap.

Reply:
Yeah, I might give this some thought and revisit this. I'm fine with this for now though.